### PR TITLE
-O4: When -O3 isn't enough

### PR DESCRIPTION
**Pull request by kripken**
This defines a new `-O4` optimization mode, as flatten + flat-only opts (currently `local-cse`) + `-O3`.

In practice, flattening is not needed for LLVM output, which is pretty flat already (no block or if values, etc., even if it does use tees and does nest expressions; and LLVM has already done gvn etc. anyhow). In general, though, wasm generated by a non-LLVM compiler may naturally be nested because wasm allows that. See for example #1593 where an AssemblyScript testcase requires flattening to be fully optimized. So `-O4` can help there.

`-O4` takes 3x longer to run than `-O3` in my testing, basically because flat IR is much bigger. But when it's useful it may be worth it. It does handle that AssemblyScript testcase and others like it. There's not much big real-world code that isn't LLVM yet, but running the fuzzer - which happily creates nested stuff all the time - I see `-O4` consistently shrink the size by around 20% over `-O3`.

Note that because the main opt levels - `-O1, -O2` etc. - match clang's, and there's no significant benefit for this opt level on LLVM output anyhow, I think this will stay only in Binaryen, that is, it wouldn't make sense to add an `-O4` to emcc.

### CODE REVIEW

File: src/tools/optimization-options.h [hunk 1]
```diff
@@ -42,27 +42,34 @@ struct OptimizationOptions : public Options {
...
                 })
-           .add("", "-O3", "execute -O3 optimization passes",
+           .add("", "-O3", "execute -O3 optimization passes (spends potentially a lot of time optimizing)",
                 Options::Arguments::Zero,
                 [this](Options*, const std::string&) {
                   passOptions.optimizeLevel = 3;
                   passOptions.shrinkLevel = 0;
                   passes.push_back(DEFAULT_OPT_PASSES);
                 })
+           .add("", "-O4", "execute -O4 optimization passes (also flatten the IR and do flat-dependent opts)",
```
Comment (jgravelle-google) [inline src/tools/optimization-options.h]: The other comments give a good heuristic to the user, whereas this one is more descriptive of what it makes Binaryen do. I'd word this more like "optimizations that usually aren't worth it, can be many times slower than -O3" to give the user a better sense of what the tradeoff is here (optimization at all costs)

**Comment by dcodeIO** [comment]
AssemblyScript currently simply passes the optimize levels down to Binaryen, and I wonder if it might make sense to run flatten + local-cse on other occasions than `-O4` as well due to the compiler usually emitting blocked code, as the improvement of 20% is quite significant.

For example, it might well be that a user wants `-Oz` with the best possible code size reduction, so does it make sense to run flatten + local-cse there as well due to AssemblyScript's design (while LLVM-based compilers wouldn't have to)?

**Comment by kripken** [comment]
@dcodeIO You can pass `--optimize-level=4 --shrink-level=2` for example to get something like `-Oz` plus the extra stuff here - and yeah, that should have benefits over `-Oz` (I didn't measure yet, though).

Or, you could manually run `--flatten --local-cse -Oz`, however, we'll probably add more passes after `--local-cse` eventually, and using `--optimize-level=4` would get them without the user needing to remember to add them.

File: src/tools/optimization-options.h (hunk 1 above)
Comment (kripken) [reply src/tools/optimization-options.h]: Good point, I'll improve this.

**Comment by dcodeIO** [comment]
Would it, maybe, make sense to have another option there like `--no-llvm`, for compilers like AssemblyScript to utilize on any optimize level, and Binaryen then doing the right thing according to achieve roughly the same at the end?

**Comment by kripken** [comment]
Interesting point. Maybe the opposite, `--llvm` that implies LLVM optimized the code already would make more sense, as there's probably a lot of variety in non-LLVM compilers, and just one LLVM? However, that does seem like a pretty specific flag...

**Comment by dcodeIO** [comment]
Yeah, not sure if there's a common denominator for what LLVM is doing before passing code to Binaryen, but if there is, maybe that'd make a good name? Maybe `--pre` for "preoptimized"?

**Comment by kripken** [comment]
Yeah, the code already being fairly optimized is an important part of this. But maybe that's captured by the difference between `-O3` and `-O4` already, though - higher levels make sense for code that needs more optimization work?

Another issue aside from being fairly optimized is how "flat" the code is - does it use block and if return values, etc. Flattening helps there. That's something LLVM doesn't do now, but might eventually.

Overall, I don't care much about the names of options here. We can add more if it helps usability.

**Comment by dcodeIO** [comment]
What I'm doing now is to [magically increase the optimize level](https://github.com/AssemblyScript/assemblyscript/commit/c6ec5e2832b3f7af9c30f47bc692a09f13de56d2) on `-O3` and `-Oz` in an attempt to keep optimization options somewhat symmetric to other toolchain (it appears to me that providing the original `-Oz`, for example, doesn't make a lot of sense without flatten etc. when AssemblyScript is the generator). But I worry that, at some point (or already), `-Oz` might deviate from `--optimizeLevel 4 --shrinkLevel 2` or something like that, so an additional option that'd allow me to just rely on Binaryen to do the right thing, no matter the optimize level, would make that more convenient for me. Not sure how valid this concern actually is, but that'd be my reasoning :)

**Comment by kripken** [comment]
@dcodeIO I don't understand, sorry. What is "other toolchain"?

**Comment by dcodeIO** [comment]
I was referring to LLVM/clang, Emscripten and Binaryen sharing optimization levels, and I thought it would be great if AssemblyScript could share as much of these as well. For example, I imagine users comparing outputs of multiple toolchains based on the same optimization levels, so unifying that seems like something that could avoid misunderstandings and feel more coherent when switching between toolchains in generaal.

An additional option indicating preoptimization by LLVM might help to achieve this, because `-O4`, as it stands, always implies all of `-O3`, even if LLVM would also flatten with `-O2` for example. I actually don't know if it does, but if it does, there'd be no way to achieve the same with the AssemblyScript compiler except specifying passes manually. For example, my understanding so far is that LLVM would always produce flattened code (again, I actually don't know), and that the AS compiler should always run flatten from `-O1` upwards.

**Pull request merged**